subsys/bluetooth/mesh/time_srv.c

```diff
@@ -504,6 +509,16 @@ int _bt_mesh_time_srv_update_handler(const struct bt_mesh_model *model)
 		return err;
 	}
 
+	/* If sent as an unsolicited message, the Time Status message shall be sent
+	 * with TTL=0 to avoid building up cumulative time errors resulting from delays
+	 * in processing the messages by relays.
+	 */
+	if (!bt_mesh_model_pub_is_retransmission(pub->mod)) {
+		srv->cached_ttl = pub->ttl ? pub->ttl : srv->cached_ttl;
+		pub->ttl = srv->is_unsolicited ? 0 : srv->cached_ttl;
+		srv->is_unsolicited = false;
+	}
```

Comment:
I'm still uncertain if this is correct. When `bt_mesh_model_publish` is called by a model, `pub->count` is set to a non-zero value:
https://github.com/zephyrproject-rtos/zephyr/blob/f9778472105d756fff7d1e5b54353421d356ed43/subsys/bluetooth/mesh/access.c#L1625
https://github.com/zephyrproject-rtos/zephyr/blob/f9778472105d756fff7d1e5b54353421d356ed43/include/zephyr/bluetooth/mesh/access.h#L680
In this case, the update callback is only called for retransmissions:
https://github.com/zephyrproject-rtos/zephyr/blob/f9778472105d756fff7d1e5b54353421d356ed43/subsys/bluetooth/mesh/access.c#L927-L953

What we probably what here is in the `bt_mesh_time_srv_time_status_send` function store the current publication ttl value, set `srv->model->pub->ttl` to zero like it was before and set the `is_unsolicited` flag. Then, when the update callback is called, `bt_mesh_model_pub_is_retransmission` returns false and `is_unsolicited` is true, we restore the old ttl and drop the `is_unsolicited` flag.

Reply:
> The update callback is unconditionally called for the first publication, always here:
> https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/access.c#L861
> over this call:
> https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/access.c#L939-L940
> exactly in `pub_period_start` function counter is configured if retransmission was configured:
> https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/access.c#L855

No, this is not correct because as I said, `pub->count` is already set. The idea here is that a model feels up the publication buffer in advance before calling `bt_mesh_model_publish`. Look at how `bt_mesh_msg_send` does this: https://github.com/zephyrproject-rtos/zephyr/blob/793c507209eb508c1c29441b56d60303df4187a7/subsys/bluetooth/mesh/msg.c#L101-L104

Reply:
Hm... I guess the line that you pointed here: https://github.com/zephyrproject-rtos/zephyr/blob/f9778472105d756fff7d1e5b54353421d356ed43/subsys/bluetooth/mesh/access.c#L1625
is access layer bug. 
If publication goes over API (not over timer) it skips the first attempt and goes directly to retransmissions.